src/modules/readout/vmm3/test/ReadoutGenerator.cpp

```diff
@@ -12,56 +12,86 @@
 #include <cstdio>
 #include <cstring>
 #include <readout/vmm3/test/ReadoutGenerator.h>
+#include <stdexcept>
 
-uint16_t ReadoutGenerator::vmm3ReadoutDataGen(
-  uint8_t *Buffer, uint16_t MaxSize, bool Randomise,
-  uint8_t Type, uint32_t SeqNum, uint8_t Rings, uint16_t NumReadouts) {
 
-  uint16_t DataSize = HeaderSize + NumReadouts * sizeof(VMM3Parser::VMM3Data);
-  if (DataSize > MaxSize) {
-    printf("Too much data for buffer. DataSize: %u, MaxSize: %u\n", DataSize,
-           MaxSize);
-    return 0;
+
+ReadoutGenerator::ReadoutGenerator(uint8_t *BufferPtr, uint16_t MaxPayloadSize,
+  uint32_t InitialSeqNum, bool Randomise)
+  : Buffer(BufferPtr)
+  , BufferSize(MaxPayloadSize)
+  , SeqNum(InitialSeqNum)
+  , Random(Randomise) { }
+
+
+uint16_t ReadoutGenerator::makePacket(uint8_t Type, uint16_t NumReadouts, uint8_t Rings) {
+  generateHeader(Type, NumReadouts);
+  generateData(Rings, NumReadouts);
+  finishPacket();
+  return DataSize;
+}
+
+void ReadoutGenerator::generateHeader(uint8_t Type, uint16_t NumReadouts) {
+
+  DataSize = HeaderSize + NumReadouts * VMM3DataSize;
+  if (DataSize >= BufferSize) {
+    throw std::runtime_error("Too many readouts for buffer size");
   }
 
-  memset(Buffer, 0, MaxSize);
-  auto DP = (uint8_t *)Buffer;
-  auto Header = (ReadoutParser::PacketHeaderV0 *)DP;
+  memset(Buffer, 0, BufferSize);
+  auto Header = (ReadoutParser::PacketHeaderV0 *)Buffer;
 
-  Header->CookieAndType = 0x535345 + (Type << 24);
+  Header->CookieAndType = (Type << 24) + 0x535345;
   Header->Padding0 = 0;
   Header->Version = 0;
   // Header->OutputQueue = 0x00;
 
   Header->TotalLength = DataSize;
   Header->SeqNum = SeqNum;
-  Header->PulseHigh = SeqNum;
-  Header->PulseLow = 10;
-  Header->PrevPulseHigh = SeqNum;
-  Header->PrevPulseLow = 0;
+  Header->PulseHigh = SeqNum; /// \todo fixme - better idea?
+  Header->PulseLow = TimeLowOffset;
+  Header->PrevPulseHigh = SeqNum; /// \todo fixme - better idea?
+  Header->PrevPulseLow = PrevTimeLowOffset;
+}
+
+
+void ReadoutGenerator::generateData(uint8_t Rings, uint16_t NumReadouts) {
+  auto DP = (uint8_t *)Buffer;
 
   DP += HeaderSize;
+  uint32_t TimeLow = TimeLowOffset + TimeToFirstReadout;
   for (auto Readout = 0; Readout < NumReadouts; Readout++) {
     auto ReadoutData = (VMM3Parser::VMM3Data *)DP;
     ReadoutData->RingId = (Readout / 10) % Rings;
+    //printf("RingId: %u\n", ReadoutData->RingId);
     ReadoutData->FENId = 0x01;
     ReadoutData->DataLength = sizeof(VMM3Parser::VMM3Data);
     assert(ReadoutData->DataLength == 20);
 
     ReadoutData->TimeHigh = SeqNum;
-    ReadoutData->TimeLow = 100 + Readout * 100;
-    ReadoutData->VMM = Readout & 0x1;
+    ReadoutData->TimeLow = TimeLow;
+    ReadoutData->VMM = Readout & 0x3;
     ReadoutData->OTADC = 1000;
-    ReadoutData->Channel = 17 + Readout % 32;
```

Comment:
Are the changes `0x1` → `0x3` and `17` → `MinChannel == 16` intentional?

Reply:
Yes both were bugs that I discovered while using this to generate an artificial dataset for Francesco.